File: src/breslin/clientside/ability/move/Normal_ProcessTick_Move.java
package breslin.clientside.ability.move;

/**********************************
*          INCLUDES
**********************************/
//math
import breslin.math.Vector3D;



/******************************************************
*				INCLUDES
********************************************************/
//parent
import breslin.clientside.ability.move.AbilityMoveState;


class Normal_ProcessTick_Move extends AbilityMoveState
{

	private static Normal_ProcessTick_Move mAbilityMoveState;

	/** A private Constructor prevents any other class from instantiating. */
	private Normal_ProcessTick_Move()
	{
		  //	 Optional Code
	}

/******************************************************
*				METHODS
*
*
*
*
********************************************************/

public static synchronized Normal_ProcessTick_Move getAbilityMoveState()
{
   	if (mAbilityMoveState == null){
    	mAbilityMoveState = new Normal_ProcessTick_Move();
    }
	    return mAbilityMoveState;
}


public Object clone()throws CloneNotSupportedException
{
    throw new CloneNotSupportedException();
}



public void enter(AbilityMove abilityMove)
{
}

public void execute(AbilityMove abilityMove)
{
	// if distance exceeds threshold && server velocity is zero
	if(abilityMove.mDeltaPosition > abilityMove.mPosInterpLimitHigh && !abilityMove.mShape.mServerCommandCurrent.mVelocity.isZero())
	{
		abilityMove.mProcessTickStateMachine.changeState(Catchup_ProcessTick_Move.getAbilityMoveState());
    }
		Vector3D serverVelocity = new Vector3D();
		serverVelocity.copyValuesFrom(abilityMove.mShape.mServerCommandCurrent.mVelocity);
		serverVelocity.normalise();

        if(abilityMove.mShape.mCommandToRunOnShape.mFrameTime != 0)
        {
			abilityMove.mShape.mSpeed = abilityMove.calcuateSpeed(
			abilityMove.mShape.mServerCommandCurrent.mVelocity,
			abilityMove.mShape.mCommandToRunOnShape.mFrameTime);
        }

		serverVelocity.multiply(abilityMove.mShape.mSpeed);

		//keep player from teleporting
		//abilityMove.regulate(serverDest);

		abilityMove.mShape.mCommandToRunOnShape.mVelocity.copyValuesFrom(serverVelocity);
}

public void exit(AbilityMove abilityMove)
{
}


}

File: src/breslin/clientside/ability/move/normal_process_tick_move.php
var NormalProcessTickMove = new Class(
{

Extends: State,
initialize: function ()
{
},

log: function(msg)
{
        setTimeout(function()
        {
                throw new Error(msg);
        }, 0);
},

enter: function(move)
{
//	this.log('enter n');
},

execute: function(move)
{
//	move.mShape.mMesh.innerHTML='n:' + move.mShape.mIndex;

        // if distance exceeds threshold && server velocity is zero
        if(move.mDeltaPosition > move.mPosInterpLimitHigh && !move.mShape.mServerCommandCurrent.mVelocity.isZero())
        {
       		move.mProcessTickStateMachine.changeState(move.mCatchupProcessTickMove);
        }
        
	serverVelocity = new Vector3D();

        serverVelocity.copyValuesFrom(move.mShape.mServerCommandCurrent.mVelocity);
	
        serverVelocity.normalise();

        if(move.mShape.mCommandToRunOnShape.mFrameTime != 0)
        {
                v = move.mShape.mServerCommandCurrent.mVelocity;
                f = move.mShape.mCommandToRunOnShape.mFrameTime;
		
                move.mShape.mSpeed = move.calculateSpeed(v,f);
        }
        serverVelocity.multiply(move.mShape.mSpeed);
 
	move.mShape.mCommandToRunOnShape.mVelocity.copyValuesFrom(serverVelocity);
	
},

exit: function(move)
{

}


});

